graphql/resolvers.go

```diff
@@ -288,6 +290,83 @@ func (r *mutationResolver) SpawnHost(ctx context.Context, spawnHostInput *SpawnH
 	return &apiHost, nil
 }
 
+func (r *mutationResolver) EditSpawnHost(ctx context.Context, editSpawnHostInput *EditSpawnHostInput) (*restModel.APIHost, error) {
+	usr := route.MustHaveUser(ctx)
+	h, err := host.FindOneByIdOrTag(editSpawnHostInput.HostID)
+	if err != nil {
+		return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error finding host by id: %s", err))
+	}
+
+	if !CanUpdateSpawnHost(h, usr) {
+		return nil, Forbidden.Send(ctx, "You are not authorized to modify this host")
+	}
+
+	if editSpawnHostInput.DisplayName != nil {
+		err = h.SetDisplayName(*editSpawnHostInput.DisplayName)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while modifying spawnhost name: %s", err))
+		}
+	}
+	if editSpawnHostInput.NoExpiration != nil {
+		if *editSpawnHostInput.NoExpiration {
+			h.MarkShouldNotExpire(ExpireInDays(evergreen.SpawnHostExpireDays))
+		} else {
+			h.MarkShouldExpire(ExpireInDays(evergreen.SpawnHostExpireDays))
+		}
+	}
+	if editSpawnHostInput.Expiration != nil {
+		err = h.SetExpirationTime(*editSpawnHostInput.Expiration)
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while modifying spawnhost expiration time: %s", err))
+		}
+	}
+	if editSpawnHostInput.InstanceType != nil {
+		config, err := evergreen.GetConfig()
+		if err != nil {
+			return nil, InternalServerError.Send(ctx, "unable to retrieve server config")
+		}
+		foundInstanceType := false
+		for _, a := range config.Providers.AWS.AllowedInstanceTypes {
+			if a == *editSpawnHostInput.InstanceType {
+				foundInstanceType = true
+				err = h.SetInstanceType(*editSpawnHostInput.InstanceType)
+				if err != nil {
+					return nil, InternalServerError.Send(ctx, fmt.Sprintf("Error while modifying spawnhost expiration time: %s", err))
+				}
+			}
+		}
+		if !foundInstanceType {
+			return nil, ResourceNotFound.Send(ctx, fmt.Sprintf("Error finding matching instance type"))
+		}
+	}
+	if editSpawnHostInput.AddedInstanceTags != nil || editSpawnHostInput.DeletedInstanceTags != nil {
+		addedTags := []host.Tag{}
+		deletedTags := []string{}
+		for _, tag := range editSpawnHostInput.AddedInstanceTags {
+			tag.CanBeModified = true
+			addedTags = append(addedTags, *tag)
+		}
+		for _, tag := range editSpawnHostInput.DeletedInstanceTags {
+			deletedTags = append(deletedTags, tag.Key)
+		}
+		opts := host.HostModifyOptions{
+			AddInstanceTags:    addedTags,
+			DeleteInstanceTags: deletedTags,
+		}
+		if err = cloud.ModifySpawnHost(ctx, evergreen.GetEnvironment(), h, opts); err != nil {
```

Comment:
why not build the `opts` variable with display name and expiration options so they don't have to be handle separately in this resolver?

Reply:
Since every field is optional i check if it exists before performing an update for that field. 
If i  group the items together theres a risk of passing in null to a field thats already populated and unintentionally changing it. I wanted to be deliberate about each change i made which is why i wrap each one in a conditional.
 Im also able to return better error messages this way since i can return errors specific to what field caused the error. Instead of returning a generic error.

Reply:
only  set variables inside `opts` will be processed/updated. all you have to do is check if the variable is nil before setting. the error messages returned from `ModifySpawnHost` will be specific to the field instead of a generic error